### ISSUE TITLE
Unwanted color change of toolbar. 

### ISSUE THREAD

**Issue by sipersso** [helpscout/beacon-ios-sdk#73]
I was testing my app for iOS 12 compatibility and found the following issue. 

Beacon: 1.0.5
iOS version: 12.4

Steps to reproduce:
1: Open the beacon (docs enabled) and close it
2: Open the beacon (docs disabled) and close it

Expected: 
The navigationbar color of my app should be unchanged. 

Actual:
The tint color of navigation bar is changed to the same color as the beacon navigationbar. 

I wasn't able to attach it here, but I do have a video to show the behavior if you want me to send it to you.

**Comment by sipersso** [comment]
After trying it again, step one isn't really necessary.  All I need to do is to open the beacon with docs disabled and then close it. If I do this, the navigationbar is changed to the helpscout tint color. I also noticed that the next time you open the beacon, the navigation bar in the beacon has the wrong color as well. 

If you look at the attached images. My screen is supposed to have the default toolbar color and the beacon is supposed to be tinted. The first time the beacon is shown, this seems to be the case, but after this it becomes like the attached images. 
<img width="344" alt="Skärmavbild 2020-06-20 kl  18 57 15" src="https://user-images.githubusercontent.com/6479433/85207283-fe506700-b327-11ea-88d1-e6bb55ad4d6a.png">
<img width="342" alt="Skärmavbild 2020-06-20 kl  18 57 40" src="https://user-images.githubusercontent.com/6479433/85207288-03151b00-b328-11ea-879a-d00e8b78badc.png">

**Comment by codenameDuchess** [comment]
Hi @sipersso, thanks again for the bug report and for tracking down how to reproduce it. Really appreciate the screenshots as well. We'll let you know when we get a fix in. 👍

**Comment by codenameDuchess** [comment]
Internal tracking ticket: https://helpscout.atlassian.net/browse/IOSSDK-203

**Comment by sipersso** [comment]
After testing I have found that I am getting this issue on iOS 13 as well. It doesn't seem to be related to iOS 12 after all.

**Comment by jotaemepereira** [comment]
> After testing I have found that I am getting this issue on iOS 13 as well. It doesn't seem to be related to iOS 12 after all.

Hey @sipersso! Thanks for all the information, we tried to reproduce this bug on iOS 13 but we weren't able to hit it. Can you provide us with some reproduction steps? Thanks 🙌 !

**Comment by sipersso** [comment]
I just tried it again, and I wasn't able reproduce it this time even after trying for 15+ minutes. So I guess we can assume that this is an iOS 12 bug. If I see it again, I can always open another issue ;)

**Comment by sipersso** [comment]
Just an update. This issue can be reproduced using the HelpScout mobile app version 1.20.3 and iOS 13

1: Open the HelpScout app
2: Press get help
3: Press get in touch
4: Close the beacon using the x button.
5: Open the beacon again by pressing get help 

Expected:
Navigationbar should have the same color (blue) after step 2 and 5

Actual:
Navigationbar color is blue in step 2, but black in step 5.

See attached video.

[toolbar_color.MP4.zip](https://github.com/helpscout/beacon-ios-sdk/files/4975506/toolbar_color.MP4.zip)

**Comment by sipersso** [comment]
Just to let you know... The "fix" in 2.0.0 has unwanted side effects.  Now the beacon looks ok, but it changes the navigation bar color for the rest of the application instead!

Steps to reproduce:
1: Open the beacon (I have docs enabled, but chat disabled)
2: Press get in touch
3: Close beacon using x button. 

Expected result:
The beacon should not affect the rest of the application

Actual result:
After opening the beacon. The color of the navigation bar is changed to whatever color the beacon uses! 

I am guessing that you are using UIAppearance to change the colors of the navigation bar (which will affect the whole app and not just your beacon) and you don't reset it? See attached video. Without a workaround or a new bug fix for this one, I am not comfortable with using this in production.

[2.0_update.mov.zip](https://github.com/helpscout/beacon-ios-sdk/files/5089402/2.0_update.mov.zip)

**Comment by sipersso** [comment]
The "get in touch" button seems to be the key here. Not sure what happens with the navigationbar there, but whenever I press "get in touch" the color will be off for all my UINavigationControllers throughout the whole app.

**Comment by sipersso** [comment]
For anyone looking for a workaround, I spent a few hours on it found the following to work. 

I have prepared two methods to configure the navigationbar appearance. One to setup the beacons style and one to reset it when done. 

```
    private func prepareBeaconAppearance(backgroundColor:UIColor, textColor:UIColor){
    
        if #available(iOS 13.0, *) {
            
            let navBarAppearance = UINavigationBarAppearance()
            navBarAppearance.configureWithDefaultBackground()
            navBarAppearance.shadowColor = nil
            navBarAppearance.backgroundColor = backgroundColor
            navBarAppearance.titleTextAttributes = [NSAttributedString.Key.foregroundColor:textColor]
            
            let navigationBarAppearace = UINavigationBar.appearance()
            navigationBarAppearace.standardAppearance = navBarAppearance
            navigationBarAppearace.scrollEdgeAppearance = navBarAppearance
            navigationBarAppearace.compactAppearance = navBarAppearance
        }
    }
    
    private func resetNavigationbarAppearance(){
        if #available(iOS 13.0, *) {
            
            let navBarApperance = UINavigationBarAppearance()
            navBarApperance.configureWithDefaultBackground()
            
            let navigationBarAppearace = UINavigationBar.appearance()
            navigationBarAppearace.scrollEdgeAppearance = navBarApperance
            navigationBarAppearace.compactAppearance = navBarApperance
            navigationBarAppearace.standardAppearance = navBarApperance
        } else {
            let navBarApperance = UINavigationBar.appearance()
            navBarApperance.backgroundColor = nil
            navBarApperance.barTintColor = nil
        }
    }

```

I call the prepareBeaconAppearance before every time I open the beacon, and resetNavigationbarApperance in the onBeaconClose callback. The first call to prepareBeaconApperance won't have any effect and the default configuration appears to work. But if you don't have this method, the beacon won't have the correct color next time you open it, so you need to do both. 

For helpscout: It is probably not a great idea to set the scrollEdgeApperance/compactAppearance/standardApperance on UINavigationBar.appearance(). You should set the apperaances on navigationController.navigationBar instead. If you do that, it won't have these side-effects.

**Comment by jotaemepereira** [comment]
Hey @sipersso, thanks for providing extra information about this issue and a workaround for this issue.

> I am guessing that you are using UIAppearance to change the colors of the navigation bar (which will affect the whole app and not just your beacon), and you don't reset it?

We do not currently change the appearance of the navigation bar from the `UINavigationBar.appeareance()`, so the issue seems trickier of what it looks. We are going to take a more in-depth look to see why this is happening; we will keep you posted.

**Comment by sipersso** [comment]
After release testing... My workaround isn't reliable and I would advise against using. It is really hard to reset the changes made on UIAppearance without side effects.

I hope there will be fix for this soon. Right now I have opted for reverting back to 1.0.5 since I cannot risk the beacon modifying the appearance of my whole app. Even though there is a bug in 1.0.5 that the navigationbar changes color for the beacon on the second time it is shown, it doesn't seem to use UIAppearance and doesn't change the appearance of the host application navigationbars.

For me, this is a show stopper issue.

**Comment by ScottPetit** [comment]
👋 Hey @sipersso,

Thanks for raising this and for providing a workaround.  This one has definitely been a doozy, but we believe we've finally resolved this in 2.0.1.  You can read more about what's in the release in our [Changelog](https://github.com/helpscout/beacon-ios-sdk/blob/master/CHANGELOG.md).  Let us know if you're still seeing issues with this and thanks again.

**Comment by andreyz** [comment]
I have exactly the same issue on iOS 14 with SDK 2.0.1. In addition the delegate callbacks are not called after second presentation #88, so I'd say that both issues should be reopened.

I'm presenting Beacon from a modally presented VC. I configure `HSBeaconSettings` in `viewDidLoad` and assign the presenting VC as a delegate. Everything works the first time (delegate callbacks and nav bar color is preserved on presenting VC). Once presenting VC is itself dismissed and later presented again (usually not second, but a third time), Beacon delegate isn't called and after Beacon dismissal the navigation bar color of the presenting VC is changed.

*Normal presenting VC appearance*
<img width="368" alt="Screenshot 2020-09-02 at 13 14 43" src="https://user-images.githubusercontent.com/205317/91974670-61924880-ed1e-11ea-8800-1299b1971a7e.png">

*Unwanted modified presenting VC appearance*
<img width="366" alt="Screenshot 2020-09-02 at 13 15 09" src="https://user-images.githubusercontent.com/205317/91974675-635c0c00-ed1e-11ea-88b6-79982bcd3a2c.png">

**Comment by sipersso** [comment]
@andreyz that is weird. I am not able to reproduce the color issue anymore with 2.0.1. Is it reproducable each time?

**Comment by andreyz** [comment]
@sipersso yes, I can reliably reproduce it in the simulator.

**Comment by andreyz** [comment]
@sipersso @ScottPetit this is probably self-inflicted, my code (due to state synchronisation issue) seems to call `HSBeacon.open` right when presenting view is being dismissed, which seems to cause the issue with the navigation bar to appear.

I'm guessing this confuses navigation bar styling code either in the SDK or iOS and applies Beacon styling to presenting view as well.

**Comment by andreyz** [comment]
Update: and the state synchronisation issue in my code is caused by `HSBeaconDelegate` callbacks not being called after second `HSBeacon.open` call in presenting VC. So the root cause is delegate not being called (it's correctly set on HSBeaconSettings, I checked).

**Comment by codenameDuchess** [comment]
Hi @andreyz thanks for the details and sorry for the trouble. We recently resolved a similar issue where the delegate function was not called in iOS 13 but it appears you are running into this issue in iOS 14. I've added a ticket for us to look into this once iOS 14 is out of beta. We'll report back with our findings.

[Internal tracking ticket](https://helpscout.atlassian.net/browse/IOSSDK-331).
